### ISSUE TITLE
[DISCUSS] More deliberate pre-release testing

### ISSUE THREAD

**Issue by alamb** [apache/datafusion#13661]
### Is your feature request related to a problem or challenge?

- Part of https://github.com/apache/datafusion/issues/13648

Up to now, when we have made DataFusion releases, we have mostly focused on validated that DataFusion's own unit tests have passed, (see [dev/release/README.md](https://github.com/apache/datafusion/blob/main/dev/release/README.md)) but haven't tested the upgrade with other downstream projects (like ballista, ray, InfluxDB IOx, etc) until *after* we have release the code

This results sometimes in downstream users finding issues after release. Some recent examples
- delta-rs upgrade: https://github.com/delta-io/delta-rs/pull/2886
- https://github.com/apache/datafusion/issues/13481
- https://github.com/apache/datafusion/issues/13568
- https://github.com/apache/datafusion/issues/13510

### Describe the solution you'd like

I would like to  improve the testing / release process for DataFusion releases to reduce the number of regressions found **after** release.

This would likely take the form of updating the [dev/release/README.md](https://github.com/apache/datafusion/blob/main/dev/release/README.md) 

### Describe alternatives you've considered

One idea mentioned by @Omega359 and @andygrove on  https://github.com/apache/datafusion/issues/13525#issuecomment-2496487538 

> Upgrading our own subprojects (Ballista, Comet, DF Python, DF Ray) as part of the DataFusion release process makes a lot of sense to validate that the upgrade guide is complete.

**Comment by alamb** [comment]
BTW I often create WIP PRs to test releases of sqlparser-rs and arow-rs with DataFusion *before* the release

For example:
- https://github.com/apache/datafusion/pull/13546

This caught at least one regression before release recently (https://github.com/apache/datafusion-sqlparser-rs/pull/1556, fixed by @goldmedal  🙏 )

**Comment by findepi** [comment]
> I would like to improve the testing / release process for DataFusion releases to reduce the number of regressions found **after** release.

I agree with the goal, but i am concerned about the cost/overhead involved. We don't have infinite bandwith at disposal.
Additionally, orchestrating testing across several downstream projects creates new problems we didn't have before

- which downstream projects can stop the release train? what kind of problems are able to stop the release train?
  - this is especially important question for closed-source downstream projects
- how long are we willing to wait for external teams to report back? Downstream project maintainers will obviously be willing and motivated to help, but their availability cannot be assumed

So what if we focused instead on:

- improving testing within DF itself? if a downstream project is concerned about stability of feature X, they can contribute to improve test coverage for feature X (eg https://github.com/apache/datafusion/issues/13470)
- easy low-overhead (automated) releases. If we see a big problem / regression after a major release, this can be patched on a maintenance branch. A maintenance branch can release daily without human intervention.

**Comment by alamb** [comment]
> I agree with the goal, but i am concerned about the cost/overhead involved. We don't have infinite bandwith at disposal.
Additionally, orchestrating testing across several downstream projects creates new problems we didn't have before

Yes, this is true. I imagine an incremental rollout type approach -- where we start with one project (datafusion-python is a natural example, and maybe we can get delta-rs to help too). In my (likely naieve) thinking the downstream projects will be willing to help as they are directly affected

**Comment by alamb** [comment]
> easy low-ceremony low-overhead (automated) releases. If we see a big problem / regression after a major release, this can be patched on a maintenance branch. A maintenance branch can release daily without human intervention

It is my understanding that the apache voting / approval process prevents automated builds

Creating a maintenance branch is also a compelling idea where we can focus on stability / shoring up test coverage 🤔

**Comment by findepi** [comment]
> It is my understanding that the apache voting / approval process prevents automated builds

That's my understanding too, but i hope this process isn't nonnegotiable. 
Processes are there to serve the project & the community after all, not the other way around.

**Comment by findepi** [comment]
That would work for me as long as these releases are the only once we publish. I would want automation for 'the releases' the people use. Especially if we have a maintenance branch, it would reasonable to release after every PR merge.
I don't know what problems the manual release process solves that cannot be solved with automated releases.

anyway, did we hijack the thread?

**Comment by alamb** [comment]
> anyway, did we hijack the thread?

Yeah, we should probably file a separate discussion about more frequent releases if we want to pursue that option

**Comment by edmondop** [comment]
@alamb I did some personal testing with Cargo mutants. We can have a manually runnable pipeline, or something scheduled every day, that could run mutation testing and inform us about the weakness of our tests 

I am happy to issue a PR if you think it's a good idea

**Comment by alamb** [comment]
> [@alamb](https://github.com/alamb) I did some personal testing with Cargo mutants. We can have a manually runnable pipeline, or something scheduled every day, that could run mutation testing and inform us about the weakness of our tests
> 
> I am happy to issue a PR if you think it's a good idea

I am not sure what mutation testing means -- I would be happy to check out a PR that had such a thing

As for running, we could add it to the "extended" test maybe: https://github.com/apache/datafusion/blob/main/.github/workflows/extended.yml

**Comment by edmondop** [comment]
I have logged [this](https://github.com/apache/datafusion/issues/14589) @alamb  there are some more details in that issue
